cardano-node/src/Cardano/CLI/Delegation/PoolRegistry.hs

```diff
@@ -0,0 +1,301 @@
+     -- We maintain the invariant that 'registry' contains only files.
+     tickers <- withExceptT pack $ ExceptT $
+       sequence <$> traverse entryTicker (selectEntries paths)
+
+     pure . Set.toList $ duplicates tickers
+     where
+       selectEntries :: [FilePath] -> [FilePath]
+       selectEntries paths =
+         flip filter paths $ (==".json") . Path.takeExtension
+
+       entryTicker :: FilePath -> IO (Either String Ticker)
+       entryTicker f =
+         (bimap (("While parsing '"<>f<>"': ")<>) sTicker) . AE.eitherDecode . BSL.fromStrict
+           <$> BS.readFile f
+
+--------------------------------------------------------------------------------
+-- * Ancillary
+--
+-- | Submission identifier is actually a 32-byte base16-encoded public key.
+--   We additionally require that the hexadecimal must be all-lowercase.
+validatePublicKey :: Text -> Either Text Id
+validatePublicKey text =
+  if      not (length s == 64 && all Char.isHexDigit s)
+  then fail $ "Public key not a 64-char-long base16: " <> take 128 s
+  else if not ((Char.toLower <$> s) == s)
+  then fail $ "Public key not all-lowercase: " <> s
+  else Right $ Id text
+  where s = unpack text
+
+duplicates :: (Ord a) => [a] -> Set a
+duplicates xs = snd $ foldl step (Set.empty, Set.empty) xs
+  where
+    step :: Ord a => (Set a, Set a) -> a -> (Set a, Set a)
+    step (seen, dups) x =
+      if Set.member x seen
+      then (seen, Set.insert x dups)
+      else (Set.insert x seen, dups)
+
+--------------------------------------------------------------------------------
+-- * Instances
+--
+instance FromJSON Submission where
+  parseJSON = AE.withObject "Submission" $ \v->
+    case validateFields v of
+      [] -> Submission
+              <$> v .: "id" <?> AE.Key "id"
+              <*> v .: "ticker" <?> AE.Key "ticker"
+              <*> v .: "cost" <?> AE.Key "cost"
+              <*> v .: "homepage" <?> AE.Key "homepate"
+              <*> v .: "margin" <?> AE.Key "margin"
+              <*> v .:? "pledge_address" <?> AE.Key "pledge_address"
+      xs -> fail $ List.unlines xs
+    where validateFields :: AE.Object -> [String]
+          validateFields v =
+            [ "Missing fields: " <> commaList missing
+            | not (Set.null missing) ] <>
+            [ "Unexpected fields: " <> commaList unexpected
+            | not (Set.null unexpected) ]
+            where
+              keys, mandatory, optional', missing, unexpected :: Set Text
+              keys = Set.fromList $ HMap.keys v
+              mandatory = Set.fromList ["id", "ticker", "cost", "homepage", "margin"]
+              optional' = Set.fromList ["pledge_address"]
+              missing = mandatory `Set.difference` keys
+              unexpected = keys `Set.difference` (mandatory <> optional')
+
+              commaList :: Set Text -> String
+              commaList = List.intercalate ", " . (show . unpack <$>) . Set.toList
```

Comment:
What about using [genericParseJSON](https://hackage.haskell.org/package/aeson-1.4.5.0/docs/Data-Aeson.html#v:genericParseJSON)? The structure looks pretty straightforward, so we could rely on generics here unless I am missing something :thinking: ?

Reply:
The choice to do manual decode is deliberate -- we want better error messages for human consumption.